### DISCUSSION TITLE
How to send an HTTP request from an event hander to the app

### DISCUSSION THREAD

**Discussion by angelhdzmultimedia** [zauberzeug/nicegui#472] [Q&A] [answered]
I was about to post a question to my issue but I found the solution and I want to share it with anyone else experiencing this while exploring nicegui (also with my forgetful future self 😂)

The issue was, I made an @ui.page('/login') with a login form, that sends a POST request to the @api.post('/api/auth/login') endpoint.

When using ThunderClient extension in VSCode to make the request, I got a successful response.

But when I made the request from the @ui.page('/login'), I got error.

Then I realized that the issue may be related to FastAPI, so I started researching in their repo.

I didn't find anything, so I made a new Python file with FastAPI related code only. 
Created an app, and served a post endpoint.

Then called a 3 seconds timer, and made a request, and bam! The infamous error again!

### And then it hit me...

When I code in Python, I usually have the server in one file, and the client in another file using  pyQt GUI, and run both files in different terminals.

But this time, I made everything in the same file, and executed that file with both the server and client code, and that's the issue! It has to do with threads...

So in pyQt when I want to do that kind of tasks, I would create a QThread and run the extra tasks in another thread.

So I separated the server and the client and bam, worked right away!

Hope I can help someone with this information.

So far I'm enjoying nicegui as a Python enthusiast for automation and as a web developer with Vue, Nuxt, Quasar.

Cheers to the team! 🤝🔥☕

**Comment by falkoschindler** [comment, marked as answer]
@angelhdzmultimedia I don't really understand the problem. Can you give a minimum example with "client and server in the same file" that does not work?

**Reply by angelhdzmultimedia** [reply]
`nice/__init__.py`
```py
# Server
@app.post('/api/auth/login', status_code=200)
async def apiLogin(loginData: LoginData):
  print(loginData)
  return loginData

# Client
@ui.page('/login', dark=True, title='Login')
def login(request: Request):
  with ui.column().classes('items-center full-width full-height justify-center'):
    ui.label('Login').classes('text-h6')
    ui.button('Login').props('color=blue').on('click', lambda: requests.post('http://0.0.0.0:8080/api/auth/login', json=dict(email='angelhdz@gmail.com', password='123456')) )
    ui.button('Back').props('color=blue').on('click', lambda: ui.open('/'))

ui.run(reload=False,dark=True)
```

**Reply by angelhdzmultimedia** [reply]
But then I made a new file, without anything related to nicegui, using FastAPI only:

`nice/test/__init__.py`

```py
from fastapi import FastAPI
import time
import requests

app = FastAPI()


@app.post('/api/auth/login')
def login():
  print('login')
  return 'login'


time.sleep(3)

requests.post('http://127.0.0.1:8000/api/auth/login', json=dict(
  email = 'angelhdz@gmail.com',
  password = '123456'
)) # Error!!!
```

Run the live server:
` uvicorn test:app --reload`

**Reply by falkoschindler** [reply]
Ok, I start to understand. I boiled it down to the following [minimum working example](https://en.wikipedia.org/wiki/Minimal_reproducible_example):

```py
import requests

from nicegui import app, ui

@app.post('/test')
def test():
    print('test...', flush=True)  # this line is not reached

@ui.page('/')
def index():
    ui.button('Test post', on_click=lambda: requests.post('http://0.0.0.0:8080/test', timeout=1))

ui.run()
```

I think the problem is that `requests.post` runs synchronously, i.e. it blocks the current thread until it receives a response. But the application can't process the request if it is blocking at the same time. So I'd suggest you to use an async HTTP library instead.

But why do you need a POST request at all? You _are_ on the server already. The whole point of NiceGUI is to implement all in the same place. The button click handler can simply do what needs to be done for the login to happen.

**Reply by rodja** [reply]
Have a look at our [auth example](https://github.com/zauberzeug/nicegui/blob/main/examples/authentication/main.py) for complete login workflow.

**Reply by angelhdzmultimedia** [reply]
> Ok, I start to understand. I boiled it down to the following [minimum working example](https://en.wikipedia.org/wiki/Minimal_reproducible_example):
> 
> ```python
> import requests
> 
> from nicegui import app, ui
> 
> @app.post('/test')
> def test():
>     print('test...', flush=True)  # this line is not reached
> 
> @ui.page('/')
> def index():
>     ui.button('Test post', on_click=lambda: requests.post('http://0.0.0.0:8080/test', timeout=1))
> 
> ui.run()
> ```
> 
> I think the problem is that `requests.post` runs synchronously, i.e. it blocks the current thread until it receives a response. But the application can't process the request if it is blocking at the same time. So I'd suggest you to use an async HTTP library instead.
> 
> But why do you need a POST request at all? You _are_ on the server already. The whole point of NiceGUI is to implement all in the same place. The button click handler can simply do what needs to be done for the login to happen.

Oh yeah!!! I read the documentation, and saw that everything can be done in the same file with the ui.page and event handlers,
but when I reached an API example with the @api decorator, that made me think I could also make a server API separate from the frontend logic but in the same file. 

Thanks to this issue, I learned many things about NiceGUI and FastAPI!

Now, time to have fun discovering all the good stuff NiceGUI brings to the table.

I have a programmer cousin that I tutored in Python, and now he seems to not want to try other languages (like TypeScript),
but thanks to NiceGUI, we can both use Vue and Quasar and share the same UI components and logic!!!

Wishing the kind and nice team health, peace, and success! 🤝🔥

**Reply by rodja** [reply]
Thank you very much :-)

> that made me think I could also make a server API separate from the frontend logic but in the same file

NiceGUI follows the philosophy of "backend first". So you "only" write backend code and reflect the state in UI elements for the user. Good software architectures still do separation of concern. But you do not do representational state transfer (eg. REST). You can just call another function, instantiate objects and do all the great stuff Python enables you to do.
